ops/testing.py

```diff
@@ -909,20 +909,25 @@ def set_leader(self, is_leader: bool = True) -> None:
         """
         self._backend._is_leader = is_leader
 
-        have_peer_relation = False
-        for relid, params in self._backend._relation_app_and_units.items():
-            if params['app'] == self._backend.app_name:
-                have_peer_relation = True
-                break
+        class Cleanuper:
+            def __init__(self, backend):
+                self._backend = backend
+
+            def __enter__(self):
+                peer_names = set(self._backend._meta.peers.keys())
+                for name in peer_names:
+                    self._backend._harness_expects_peer_relation.append(name)
+                    break
+                return self
+
+            def __exit__(self, exception_type, exception_vaue, exception_traceback):
+                self._backend._harness_expects_peer_relation = []
 
         # Note: jam 2020-03-01 currently is_leader is cached at the ModelBackend level, not in
         # the Model objects, so this automatically gets noticed.
-        if self._backend._is_leader and self._charm is not None and self._hooks_enabled:
-            if not have_peer_relation:
-                raise RuntimeError(
-                    'cannot trigger leader-elected without an existing peer relation: '
-                    'use Harness.add_relation(<peer-relation>, <app>) before Harness.set_leader')
-            self._charm.on.leader_elected.emit()
+        if is_leader and self._charm is not None and self._hooks_enabled:
+            with Cleanuper(self._backend):
+                self._charm.on.leader_elected.emit()
```

Comment:
Ok a new idea:
Harness could patch `leader_elected.emit()` to check whether any peer relations exist.

In harness.init:

```python
def wrap(event_handler, backend):
    def wrapper():
        assert backend.meta.peers.keys(), "cannot trigger leader-elected without..."
        event_handler()
    return wrapper
self._charm.on.leader_elected = wrap(self._charm.on.leader_elected, self._charm.model.backend)
```